src/uct/ib/mlx5/gga/gga_mlx5.c

```diff
@@ -854,16 +1034,60 @@ UCT_TL_DEFINE_ENTRY(&uct_gga_component, gga_mlx5, uct_gga_mlx5_query_tl_devices,
                     uct_gga_mlx5_iface_config_table,
                     uct_gga_mlx5_iface_config_t);
 
-UCT_SINGLE_TL_INIT(&uct_gga_component, gga_mlx5, ctor,,)
+UCT_SINGLE_TL_INIT(&uct_gga_component, gga_mlx5, ctor,
+                   kh_init_inplace(resolved_rkeys, &uct_gga_mlx5_rkey_cache),
+                   kh_destroy_inplace(resolved_rkeys, &uct_gga_mlx5_rkey_cache))
 
 static void uct_ib_mlx5_gga_md_close(uct_md_h md)
 {
     uct_gga_mlx5_md_t *gga_md = ucs_derived_of(md, uct_gga_mlx5_md_t);
+    uct_gga_mlx5_rkey_hash_key_t hash_key;
+    uct_gga_mlx5_rkey_bundle_t hash_val;
+    khash_t(garbage_rkeys) grkeys;
+    khint_t iter;
+    ucs_kh_put_t put_ret UCS_V_UNUSED;
+
+    ucs_trace("md %p: closing on GGA device %s", md,
+              uct_ib_device_name(&gga_md->super.super.dev));
+
+    kh_init_inplace(garbage_rkeys, &grkeys);
+    pthread_mutex_lock(&uct_gga_mlx5_rkeys_lock);
+    kh_foreach(&uct_gga_mlx5_rkey_cache, hash_key, hash_val, {
+        if (gga_md == hash_key.md) {
+            uct_gga_mlx5_md_rkey_handle_release(hash_key.md, &hash_val);
+            kh_put(garbage_rkeys, &grkeys, hash_key, &put_ret);
```

Comment:
You iterate through keys of the hash map, and put keys to hash set with the same hash function. It means that you don't need hash set here. It looks like an array of hash map iterators is enough.

Reply:
It's needed to safely cleanup the hash map by keys. Are you sure that one iterator of khash remains valid after deleting other one?

Reply:
I see. Then it should an array of keys.

Reply:
I used `khash` in set mode to simpify logic with (re-)allocation, we can use `ucs_array` but I dont see much benefit on cleanup path.

Reply:
The point is to use the simplest data structure that satisfies the conditions of the problem. You do not need the features provided by the set. You need only an array. You can use dynamic array for the purpose. You also can get rid of the extra code hidden by the macro in the beginning of the file.

Reply:
dynamic array with grow factor  2 may consume more memory for example, my point is both containers have pros and cons

Reply:
khash set also does reallocation in frames of rehashing when the size changes. dynamic array is more lightweight акщь this perspective. I don't think it's an advantage from this point of view.

Reply:
> khash also does reallocation in frames of rehashing when the size changes

yes but by bucket size vs miltiplying by 2 for array, so it's hard to say what's defenitely better

Reply:
ucs_list requires additional type/field for an entry

Reply:
You added a structure with redundant fields along with 10 extra methods to introduce khash set. This is an extra stuff which is not required for your purpose.

Reply:
Adding list entry to the key structure means we need to store extra 8 bytes per key, adding new inherited structure means extra allocation on cleanup every key to add the key to the list.

Reply:
If we characterize them in terms of complexity they are the same.
Agree that ucs_list (double linked list) is not needed here since you do not need to insert in the middle.
So ucs_array is preferrable solution then, a typical choice for a simple variable-length collection.

>> khash also does reallocation in frames of rehashing when the size changes

> yes but by bucket size vs miltiplying by 2 for array, so it's hard to say what's defenitely better

khash does the same. So `ucs_array` is better.

Reply:
In current state of the PR this discussion is meaningless since we need to decide if we would like to have global hash table or separate it per MD, in second case this code is not relevant at all. So, for now I propose keep the khash just because it's already implemented and work as expected. In case of decision that we go with "global hash table", I agree that ucs_array is preferable as a local storage.